src/python/pants/backend/python/target_types_rules.py

```diff
@@ -143,29 +153,160 @@ async def inject_pex_binary_entry_point_dependency(
+                See {url}.
+                """
+                )
+            )
+
+        entry_points[section][name] = entry_point
+
+    return ResolvedPythonDistributionEntryPoints(
+        FrozenDict({key: FrozenDict(value) for key, value in entry_points.items()})
+    )
+
+
 class InjectPythonDistributionDependencies(InjectDependenciesRequest):
     inject_for = PythonDistributionDependencies
 
 
 @rule
 async def inject_python_distribution_dependencies(
-    request: InjectPythonDistributionDependencies,
+    request: InjectPythonDistributionDependencies, python_infer_subsystem: PythonInferSubsystem
 ) -> InjectedDependencies:
-    """Inject any `.with_binaries()` values, as it would be redundant to have to include in the
-    `dependencies` field."""
-    original_tgt = await Get(WrappedTarget, Address, request.dependencies_field.address)
-    with_binaries = original_tgt.target[PythonProvidesField].value.binaries
-    if not with_binaries:
+    """Inject dependencies that we can infer from entry points in the distribution."""
+    if not python_infer_subsystem.entry_points:
         return InjectedDependencies()
-    # Note that we don't validate that these are all `pex_binary` targets; we don't care about
-    # that here. `setup_py.py` will do that validation.
-    addresses = await Get(
-        Addresses,
-        UnparsedAddressInputs(
-            with_binaries.values(), owning_address=request.dependencies_field.address
+
+    original_tgt = await Get(WrappedTarget, Address, request.dependencies_field.address)
+    explicitly_provided_deps, entry_points = await MultiGet(
+        Get(ExplicitlyProvidedDependencies, DependenciesRequest(original_tgt.target[Dependencies])),
+        Get(
+            ResolvedPythonDistributionEntryPoints,
+            ResolvePythonDistributionEntryPointsRequest(
+                original_tgt.target[PythonDistributionEntryPoints]
+            ),
         ),
     )
-    return InjectedDependencies(addresses)
+
+    entry_point_addresses = Addresses()
+
+    if entry_points.val:
+        address = original_tgt.target.address
+        all_entry_points = [
+            (key, name, entry_point)
+            for key, section in entry_points.val.items()
+            for name, entry_point in section.items()
+        ]
+        all_owners = await MultiGet(
+            Get(PythonModuleOwners, PythonModule(entry_point.module))
+            for _, _, entry_point in all_entry_points
+        )
+        for (key, name, entry_point), owners in zip(all_entry_points, all_owners):
+            field_str = repr({key: {name: entry_point.spec}})
+            explicitly_provided_deps.maybe_warn_of_ambiguous_dependency_inference(
+                owners.ambiguous,
+                address,
+                import_reference="module",
+                context=(
+                    f"The distribution target {address} has the field "
+                    f"`entry_points={field_str}`"
```

Comment:
Even better is something like:

https://github.com/pantsbuild/pants/blob/7362f6a336845c1c987324def19aee3179364a00/src/python/pants/backend/python/target_types_rules.py#L126-L130

Reply:
They result in the same, except I didn't have the last part to which py module it mapped. Added.